File: sdk/python/fluxcast/fcdp.py
"""Strict FCDP v0.1 header encoding and decoding."""
from dataclasses import dataclass
import struct

HEADER_LEN = 37
MAX_DATAGRAM = 1200

def _crc16(data: bytes) -> int:
    value = 0xFFFF
    for byte in data:
        value ^= byte << 8
        for _ in range(8):
            value = ((value << 1) ^ 0x1021) & 0xFFFF if value & 0x8000 else (value << 1) & 0xFFFF
    return value

@dataclass(frozen=True)
class FcdpHeader:
    packet_type: int = 3
    flags: int = 0
    session_id: int = 1
    stream_id: int = 1
    epoch: int = 0
    sequence: int = 1
    frame_id: int = 1
    fragment_index: int = 0
    fragment_count: int = 1
    priority: int = 0
    deadline_ms: int = 1000

    def _without_crc(self, payload_length: int) -> bytes:
        if not 0 <= self.priority <= 3 or not 0 <= self.fragment_index < self.fragment_count:
            raise ValueError("invalid FCDP priority or fragment range")
        return b"FC" + bytes((1, self.packet_type, self.flags, 0)) + struct.pack(
            ">QHHIIHHBHH", self.session_id, self.stream_id, self.epoch, self.sequence,
            self.frame_id, self.fragment_index, self.fragment_count, self.priority,
            self.deadline_ms, payload_length)

def encode(header: FcdpHeader, payload: bytes) -> bytes:
    if len(payload) + HEADER_LEN > MAX_DATAGRAM:
        raise ValueError("FCDP datagram exceeds 1200-byte budget")
    raw = header._without_crc(len(payload))
    return raw + struct.pack(">H", _crc16(raw)) + payload

def decode(packet: bytes) -> tuple[FcdpHeader, bytes]:
    if len(packet) < HEADER_LEN or packet[:2] != b"FC" or packet[2] != 1:
        raise ValueError("invalid FCDP header")
    raw = packet[:35]
    if _crc16(raw) != struct.unpack(">H", packet[35:37])[0]:
        raise ValueError("invalid FCDP header CRC")
    values = struct.unpack(">QHHIIHHBHH", raw[6:])
    payload = packet[37:]
    if values[-1] != len(payload): raise ValueError("FCDP payload length mismatch")
    header = FcdpHeader(packet_type=raw[3], flags=raw[4], session_id=values[0], stream_id=values[1], epoch=values[2], sequence=values[3], frame_id=values[4], fragment_index=values[5], fragment_count=values[6], priority=values[7], deadline_ms=values[8])
    header._without_crc(len(payload))
    return header, payload

File: sdk/go/fcdp/fcdp.go
// Package fcdp implements strict FCDP v0.1 packet framing.
package fcdp

import (
	"encoding/binary"
	"errors"
)

const HeaderLen = 37
const MaxDatagram = 1200

type Header struct {
	PacketType, Flags            uint8
	SessionID                    uint64
	StreamID, Epoch              uint16
	Sequence, FrameID            uint32
	FragmentIndex, FragmentCount uint16
	Priority                     uint8
	DeadlineMS                   uint16
}

func crc16(b []byte) uint16 {
	v := uint16(0xffff)
	for _, x := range b {
		v ^= uint16(x) << 8
		for i := 0; i < 8; i++ {
			if v&0x8000 != 0 {
				v = (v << 1) ^ 0x1021
			} else {
				v <<= 1
			}
		}
	}
	return v
}
func Encode(h Header, payload []byte) ([]byte, error) {
	if len(payload)+HeaderLen > MaxDatagram {
		return nil, errors.New("FCDP datagram exceeds 1200-byte budget")
	}
	if h.Priority > 3 || h.FragmentCount == 0 || h.FragmentIndex >= h.FragmentCount {
		return nil, errors.New("invalid FCDP fragment")
	}
	p := make([]byte, HeaderLen+len(payload))
	copy(p, "FC")
	p[2] = 1
	p[3] = h.PacketType
	p[4] = h.Flags
	binary.BigEndian.PutUint64(p[6:], h.SessionID)
	binary.BigEndian.PutUint16(p[14:], h.StreamID)
	binary.BigEndian.PutUint16(p[16:], h.Epoch)
	binary.BigEndian.PutUint32(p[18:], h.Sequence)
	binary.BigEndian.PutUint32(p[22:], h.FrameID)
	binary.BigEndian.PutUint16(p[26:], h.FragmentIndex)
	binary.BigEndian.PutUint16(p[28:], h.FragmentCount)
	p[30] = h.Priority
	binary.BigEndian.PutUint16(p[31:], h.DeadlineMS)
	binary.BigEndian.PutUint16(p[33:], uint16(len(payload)))
	binary.BigEndian.PutUint16(p[35:], crc16(p[:35]))
	copy(p[37:], payload)
	return p, nil
}
func Decode(p []byte) (Header, []byte, error) {
	var h Header
	if len(p) < HeaderLen || string(p[:2]) != "FC" || p[2] != 1 {
		return h, nil, errors.New("invalid FCDP header")
	}
	if binary.BigEndian.Uint16(p[35:]) != crc16(p[:35]) {
		return h, nil, errors.New("invalid FCDP CRC")
	}
	n := int(binary.BigEndian.Uint16(p[33:]))
	if len(p)-HeaderLen != n {
		return h, nil, errors.New("FCDP payload length mismatch")
	}
	h = Header{p[3], p[4], binary.BigEndian.Uint64(p[6:]), binary.BigEndian.Uint16(p[14:]), binary.BigEndian.Uint16(p[16:]), binary.BigEndian.Uint32(p[18:]), binary.BigEndian.Uint32(p[22:]), binary.BigEndian.Uint16(p[26:]), binary.BigEndian.Uint16(p[28:]), p[30], binary.BigEndian.Uint16(p[31:])}
	if h.Priority > 3 || h.FragmentCount == 0 || h.FragmentIndex >= h.FragmentCount {
		return h, nil, errors.New("invalid FCDP fragment")
	}
	return h, p[37:], nil
}
